### DISCUSSION TITLE
VSBrowser quit and start methods

### DISCUSSION THREAD

**Discussion by Andrei10Toma** [redhat-developer/vscode-extension-tester#1246] [Q&A] [answered]
Hello!
I have been using vscode-extension-tester for a while in the last year and a half and we integrated it into our VS Code extension development process. We find it very helpful for testing our extension!

We saw that VSBrowser has two methods: quit and start. We decided to use these methods to make our tests independent of each other. The "after" tests will call VSBrowser.instance.quit(), and then it will call VSBrowser.instance.start(codeBin). We observed that start method will launch a fresh VS Code window, ensuring us that there are no side effects from the previous tests.

Looking through the process trees I see that first the process tree of VS Code looks like this:
![image](https://github.com/redhat-developer/vscode-extension-tester/assets/57085701/06fa2db8-6d0c-4705-b944-d0c9d8fca26a)
Everything is launched under the chromedriver process.

After calling a quit and a start (from the "after" of the test) the process tree will look something like this:
![image](https://github.com/redhat-developer/vscode-extension-tester/assets/57085701/b8b1863f-6ac5-43e7-ac87-8d526351d89b)
As you see only one VS Code process will remain under the chromedriver and the rest of the VS Code processes will start under the init process.

Is it ok to be like this?
Is it safe to call the quit and start functions so often?

**Comment by Andrei10Toma** [comment]
Another thing that I observed is that sometimes the chromedriver process detaches the VS Code process and will be under init.
![image](https://github.com/redhat-developer/vscode-extension-tester/assets/57085701/f259568b-0c26-4425-b592-055deb149fd8)

When this happens the entire test suite just freezes and the only option to stop the process is to kill it.
Not even the timeout from the .mocharc.js is not working.
Do you have any idea what could cause it?

**Comment by caucbianca** [comment]
Hi, is there any update on this?

**Comment by djelinek** [comment, marked as answer]
Hello, thank you for bringing up this topic.

Use case you have described is not tested. It is not recommended approach how to use ExTester API. It is possible but I would describe it as experimental option at the moment.

So, unfortunately I do not have clear simple answer for you on this topic.

I can imagine this approach can bring some issues and I am not sure how efficient it can be at the end. 

The thing is that ExTester will use `start` method during calling `runTests` from an API in the `beforeAll` called for all tests you have picked to run.  In the end there is `quit` method called during `afterAll` in same manner.

If you will call `quit` and `start` multiple times between tests I would expect that the process tree could be changed.. corrupted or mixed in some way.

---
closing with some inline answers for your quoted questions

"Is it ok to be like this?"
- not sure, but it is not forbidden at the moment

"Is it safe to call the quit and start functions so often?"
- cannot say, it is not tested, in experimental manner it can work

"Do you have any idea what could cause it?"
- not sure, but it can be problem between chromedriver and VS Code electron app testing instance has lost sync. and there is also mocha run on top of it
